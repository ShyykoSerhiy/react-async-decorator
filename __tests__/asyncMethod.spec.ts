import * as React from 'react';
import { renderToString } from 'react-dom/server'
import { createApi, IApi } from './helpers';
import { asyncMethod, createFetcher, Fetcher } from '../index';

class Test extends React.Component<{ fetcher: Fetcher<string>}, {}> {
    render() {
        return React.createElement('div', {}, this.renderComponent());
    }
    @asyncMethod
    renderComponent() {
        const data = this.props.fetcher.get();
        return React.createElement('div', {}, data);
    }
    renderLoader() {
        return React.createElement('div', {}, 'loading');
    }
    renderError(err: Error) {
        return React.createElement('div', {}, err.message);
    }
}

it('render loading', () => {
    const api = createApi<string>();
    const fetcher = createFetcher(api.fetch);
    const component = renderToString(React.createElement(Test, { fetcher }))
    expect(component).toMatchSnapshot();
});

it('render error', async () => {
    const api = createApi<string>();
    const fetcher = createFetcher(api.fetch);
    api.reject(new Error('custom error'));
    try {
        fetcher.get();
        expect(false).toBeTruthy()
    } catch (e) {
        expect(true).toBeTruthy();
    }
    await api.defer.catch(_ => Promise.resolve());
    const component = renderToString(React.createElement(Test, { fetcher }))
    expect(component).toMatchSnapshot();
});

it('render data', async () => {
    const api = createApi<string>();
    const fetcher = createFetcher(api.fetch);
    api.resolve('content');
    try {
        fetcher.get();
        expect(false).toBeTruthy()
    } catch (e) {
        expect(true).toBeTruthy();
    }
    await api.defer;
    const component = renderToString(React.createElement(Test, { fetcher }))
    expect(component).toMatchSnapshot();
});
