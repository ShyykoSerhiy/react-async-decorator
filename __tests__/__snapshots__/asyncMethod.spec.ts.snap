// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render data 1`] = `"<div><div>content</div></div>"`;

exports[`render error 1`] = `"<div data-reactroot=\\"\\"><div>custom error</div></div>"`;

exports[`render loading 1`] = `"<div data-reactroot=\\"\\"><div>loading</div></div>"`;

exports[`render multiple components data 1`] = `"<div><div>content</div></div>"`;

exports[`render multiple components data 2`] = `"<div><div>content</div></div>"`;

exports[`render multiple components data 3`] = `"<div><div>content</div></div>"`;
